lib/cn/kblock_builder.c

```diff
@@ -422,11 +423,10 @@ hash_set_reset(struct hash_set *hs)
 void
 hash_set_free(struct hash_set *hs)
 {
-    struct hash_set_part *part;
-    struct hash_set_part *tmp;
+    struct hash_set_part *part, *next;
 
-    list_for_each_entry_safe (part, tmp, &hs->part_list, part_link)
-        free(part);
+    list_for_each_entry_safe(part, next, &hs->part_list, part_link)
+        vlb_free(part, sizeof(*part) + sizeof(part->hashvec[0]) * part->n_maxused);
```

Comment:
`vlb_free` says:
```
 * %used must be the exact size of the allocation given to vlb_alloc()
 * if it was larger than VLB_ALLOCSZ_MAX.  Otherwise, it should be
 * a best effort estimate of how much of the buffer was actually
 * modified.
```
Is that accurate?  If so, then why not pass in `VLB_ALLOCSZ_MAX` ?

Reply:
Because vlb only caches the first 2M of the buffer and tosses the remainder.

Back up.  The vlb cache in total pages is yay big, but we desire to cache lots of partially populated buffers because the major expense of large allocations is inserting them into the address space (whereas the pages are added to it gradually).  It's that lousy rw_sem lock IIRC used to modify the address space that really kills perf.

So while we keep a bunch of 4B buffers lying around for fast allocation, they are only half populated with pages (at best) to avoid having a ton of wired down memory.  The 2M "keep size" was chosen to accomodate the various compaction buffers, which never use more than that, so each buffer used for that purpose is fully populated.

So, when freeing if we tell vlb_free() we used 2M or less then it just caches the buffer.  Otherwise, it first does a cheap MADV_DONTNEED on the pages in excess of 2M then caches the buffer.  This both maximize the number of buffer we keep on hand, and minimizes their wired-down footprint.

So, worst case, you touch all 4M-4096 bytes in the allocation, then tell vlb_free() you only touched 1 byte.  The footprint of the cache will then have 2M more in it than it thinks, and this could get onerous on a small memory machine.  So we endeavor to tell vlb_free() exacly how much memory in the allocation that we touched.

Reply:
I misunderstood the `vlb_free` comment. My bad.